### ISSUE TITLE
Type-checked loop conditions

### ISSUE THREAD

**Issue by Rudxain** [microsoft/TypeScript#51036]
# Suggestion

There should be a `tsconfig` option to enable strict type-checking of loop condition predicates

## 🔍 Search Terms

while, loop, for, bool, boolean

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

A `tsconfig` bool option named `typeCheckedLoops` (or something else) that throws an error when any condition isn't a `boolean`

## 📃 Motivating Example

I was surprised that TS allowed the following (even with very strict config):
```ts
while ("true");

while (!"");

while ([]);

for (;{};);
```

None of those are bools, yet they don't trigger any errors

## 💻 Use Cases

This becomes a problem when the condition is a variable, or an expression composed of multiple vars.

**Comment by MartinJohns** [comment]
Essentially a duplicate of #38826.

**Comment by Rudxain** [comment]
Thank you for the info. I'll close this
